### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by magdapoppins**
- Run formatter
- (Suggestion) Add github action to automatically lint on push

### CODE REVIEW

**Comment by maxpumperla** [comment]
@magdapoppins we already use `black` for linting, so this is somewhat superfluous. @marctorsoc any preference regarding GitHub actions over pre-commit hooks?

**Comment by marctorsoc** [comment]
what formatter did you run @magdapoppins ?

I'm not familiar with Github Actions. Is this supposed to run only when we merge the PR? or also when someone pushes to a PR?

**Comment by magdapoppins** [comment]
> @magdapoppins we already use `black` for linting, so this is somewhat superfluous. @marctorsoc any preference regarding GitHub actions over pre-commit hooks?

I ran black, apparently since it is only recommended and not enforced there were some unlinted parts? Or did I somehow run it the wrong way?

Actually it would probably be better to just add linting to the travis build? This action would just run the linter but not apply the changes so any errors could be seen in a pr. This could be a way of making sure people remember to lint before a pr is merged.

**Comment by maxpumperla** [comment]
@magdapoppins running the linter in PRs would be fantastic.

**Comment by marctorsoc** [comment]
> > @magdapoppins we already use `black` for linting, so this is somewhat superfluous. @marctorsoc any preference regarding GitHub actions over pre-commit hooks?
> 
> I ran black, apparently since it is only recommended and not enforced there were some unlinted parts? Or did I somehow run it the wrong way?
> 
> Actually it would probably be better to just add linting to the travis build? This action would just run the linter but not apply the changes so any errors could be seen in a pr. This could be a way of making sure people remember to lint before a pr is merged.
> 
> I also notice the build isn't passing due to a timeout, has it been passing before these changes? I can look into it more, but let me know if this failure is expected!

I'm surprised that black just changed that, but it's really good news then! yes, let's add a lint step in travis, so we have a pipeline with two steps in CI: "lint" and "test". Eventually would be great to add mypy, flake8 and isort, but that's for another story

I would remove the github actions for now, don't see the benefit if we have the "lint" step explained above, and I don't like the idea of people pushing stuff and appearing differently in the PR

The tests are flaky, so might be as easy as pushing an empty commit and run again. The one that failed was 2.7, which we are trying to get rid of, so not your fault at all! @maxpumperla would be great if you take a look at https://github.com/hyperopt/hyperopt/issues/744

**Comment by magdapoppins** [comment]
@marctorsoc I'm quite new to Travis, not sure how to make separate steps but I did some polish in the file and added `black .` in the steps there, could we start with that? If I understand correctly we would otherwise need to create another pipeline for that linting?

File: .travis.yml
Comment (cclauss) [inline .travis.yml:48]: Given that pip now has a real dependency resolver, would it be better to have even few pip commands with more dependencies in each command?

Also, Python 3.5 is end-of-life and Python 3.8 and 3.9 are not currently being tested.

**Review by cclauss — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (marctorsoc) [reply .travis.yml:48]: we're moving to python >= 3.7, see my other PR https://github.com/hyperopt/hyperopt/pull/742. I think 3.7 and 3.8 is enough for now, let's do 3.9 when it's a bit more mature

re pip: let's tackle that in a separate PR but I like the idea of having a setup.py and be it dependant on requirements.txt files. Or we could use something like piptools  https://github.com/jazzband/pip-tools

**Comment by marctorsoc** [comment]
> @marctorsoc I'm quite new to Travis, not sure how to make separate steps but I did some polish in the file and added `black .` in the steps there, could we start with that? If I understand correctly we would otherwise need to create another pipeline for that linting?

you won't be able to use black if python2 is being tested. Let's wait until my other PR is merged https://github.com/hyperopt/hyperopt/pull/742, so that we only test python>=3.7

File: .travis.yml
Comment (marctorsoc) [reply .travis.yml:48]: regarding changes in this PR, I'm happy with something more compact, so LGTM

**Comment by marctorsoc** [comment]
> @marctorsoc I'm quite new to Travis, not sure how to make separate steps but I did some polish in the file and added `black .` in the steps there, could we start with that? If I understand correctly we would otherwise need to create another pipeline for that linting?

if you want to move to CircleCI, I'm happy to support that (in a separate PR). They have support for OS projects

**Comment by cclauss** [comment]
GitHub Actions might be a more modern/flexible/builtin choice than Travis CI or CircleCI.

**Comment by marctorsoc** [comment]
> GitHub Actions might be a more modern/flexible/builtin choice than Travis CI or CircleCI.

according to this https://blogs.vmware.com/opensource/2020/04/02/ci-tests-tools/#:~:text=A%20benefit%20for%20GitHub%20Actions,tools%20to%20help%20with%20this., there are some pros and cons of each option. If you want to go with GHA because you know more about it, sounds good to me

**Comment by magdapoppins** [comment]
Added the suggested build stages to travis as per the link @marctorsoc shared, and checked that the file is now valid according to at least this online test tool: https://config.travis-ci.com/explore - I needed to move from the matrix to a list of jobs but added those same python version tests.

I think travis is good since we already have it, no need to spend time moving to GH actions or CircleCI as long as it does what it's supposed to, right? 

There are some conflicts, I can fix those (or if anyone else wants to take over this branch, feel free) once I get some thumbs-up on this and #742 is done?

**Comment by maxpumperla** [comment]
agree on @magdapoppins 's view on being pragmatic here. Long term I can see how having all tooling "in-house" from GitHub might be very beneficial. I never cared much for the, frankly, tiny differences between circleci, travis and other incumbents. Those tools predictably converge to the same functionality anyway.

Having said that, I'd really love to play around with GitHub actions at some point, maybe another PR?

**Comment by marctorsoc** [comment]
> agree on @magdapoppins 's view on being pragmatic here. Long term I can see how having all tooling "in-house" from GitHub might be very beneficial. I never cared much for the, frankly, tiny differences between circleci, travis and other incumbents. Those tools predictably converge to the same functionality anyway.
> 
> Having said that, I'd really love to play around with GitHub actions at some point, maybe another PR?

Agreed. If someone wants to play with it in another PR, that'd be great

**Comment by marctorsoc** [comment]
> Added the suggested build stages to travis as per the link @marctorsoc shared, and checked that the file is now valid according to at least this online test tool: https://config.travis-ci.com/explore - I needed to move from the matrix to a list of jobs but added those same python version tests.
> 
> I think travis is good since we already have it, no need to spend time moving to GH actions or CircleCI as long as it does what it's supposed to, right?
> 
> There are some conflicts, I can fix those (or if anyone else wants to take over this branch, feel free) once I get some thumbs-up on this and #742 is done?

Thanks @magdapoppins. The PR is now merged. Please solve conflicts. You should be able to simplify  what you have :)

File: .travis.yml
Comment (cclauss) [inline .travis.yml:4]: Please move this above line 1 because the `os` and the `dist` are directly related.

File: .travis.yml
Comment (cclauss) [inline .travis.yml:30]: I think we can lose nose if we are running pytest.
```suggestion
```

**Review by cclauss — APPROVED** [review]
Looking good!

File: .travis.yml
Comment (marctorsoc) [reply .travis.yml:4]: it is easy / free to test on mac? would be nice to test at least one python version

File: .travis.yml
Comment (marctorsoc) [reply .travis.yml:30]: there are some tests in each style, so nose is needed as it's imported

File: .travis.yml
Comment (marctorsoc) [inline .travis.yml:40]: we don't need this anymore. They all use the same

File: .travis.yml
Comment (marctorsoc) [inline .travis.yml:46]: same here

File: .travis.yml
Comment (marctorsoc) [inline .travis.yml:52]: and here

**Review by marctorsoc — CHANGES_REQUESTED** [review]
Great job! please remove the unnecessary env variables and it's good to go

**Review by marctorsoc — APPROVED** [review]

**Comment by marctorsoc** [comment]
thanks @magdapoppins

**Pull request merged**
